nmpolicy/internal/resolver/helper.go

```diff
@@ -0,0 +1,87 @@
+package resolver
+
+import (
+	"fmt"
+
+	"github.com/nmstate/nmpolicy/nmpolicy/internal/ast"
+)
+
+func applyFuncOnPath(inputState interface{}, path []ast.Node, filterFn func(map[string]interface{}, string) (interface{}, error), shouldFilterSlice bool) (interface{}, error) {
+	if len(path) == 0 {
+		return inputState, nil
+	}
+	toFilterMap, isMap := inputState.(map[string]interface{})
+	if len(path) == 1 && isMap {
+		if filterFn != nil {
+			key := *path[0].Identity
+			outputState, err := filterFn(toFilterMap, key)
+			if err != nil {
+				return nil, err
+			}
+			return outputState, nil
+		}
+		return inputState, nil
+	} else {
+		var currentStep ast.Node
+		var nextPath []ast.Node
+
+		toFilterSlice, isSlice := inputState.([]interface{})
+		if isSlice {
+			currentStep = path[0]
+			nextPath = path
+		} else if len(path) > 1 {
+			currentStep = path[0]
+			nextPath = path[1:]
+		} else {
+			return nil, fmt.Errorf("invalid path: invalid type %T for identity step '%s'", inputState, *path[0].Identity)
+		}
+
+		if currentStep.Identity != nil {
+			key := *currentStep.Identity
+
+			if isMap {
+				value, ok := toFilterMap[key]
+				if ok {
+					value, err := applyFuncOnPath(value, nextPath, filterFn, shouldFilterSlice)
+					if err != nil {
+						return nil, err
+					}
+					if value != nil {
+						return map[string]interface{}{key: value}, nil
+					}
+					return nil, nil
+				} else {
+					return nil, fmt.Errorf("invalid path: key %s doesnt exist in %v", key, inputState)
+				}
+			}
+
+			if isSlice {
+				filteredSlice := []interface{}{}
+				for _, valueToCheck := range toFilterSlice {
+					value, err := applyFuncOnPath(valueToCheck, nextPath, filterFn, false)
+					if err != nil {
+						return nil, err
+					}
+					if value != nil {
+						filteredSlice = append(filteredSlice, valueToCheck)
+					}
+				}
+				if len(filteredSlice) > 0 {
+					if shouldFilterSlice {
+						return filteredSlice, nil
+					}
+					return toFilterSlice, nil
+				}
+				return nil, nil
+			}
+
+			return nil, fmt.Errorf("invalid path: invalid type %T for identity step '%s'", inputState, key)
+		} else {
+			return nil, fmt.Errorf("invalid path: %v has unsupported fromat", currentStep)
+		}
+	}
+}
+
+func strPtr(str string) *string {
+	return &str
+}
```

Comment:
Move this type of functions to `pointer.go`

Reply:
removed it. It was a leftover.